### PR TITLE
Feat/tokens

### LINKED ISSUE #826: Token composition utility
## User story
As a site builder or brand library creator
I want to be able to create and compose tokens with metadata
So that they can be easily organized and catalogued

## Acceptance Criteria
AC1.  @bodiless/fclasses exports an `asToken` utility which can be used to create or compose tokens with metadata.
AC2. asToken accepts a list of HOC's and returns an HOC which is equivalent to flow(...hocs)
AC3. Any static properties attached to the original component passed to this HOC are preseved.
AC4. If any of the composed HOC's adds or alters any static properties, these additions or alterations persist on the final component.
AC5. It is possible to add metadata directly to the token composed via asToken.  Such metadata will be aggregated and applied to the final component.
AC6. It is possible to add a filter to a token composed via asToken.  Such a filter will be applied to constituent tokens which precede it in the composition, as well as to any tokens (and their constituents) anoto which this token is later composed.  Filtered tokens which do not meet the filter criteria will not be applied to the final component.

## Notes
See [this spec](https://github.com/wodenx/Bodiless-JS/blob/doc/post-pilot/packages/fclasses/doc/Refactor.md#tokens-httpsgithubcomjohnsonandjohnsonbodiless-jsissues826)

### PR DESCRIPTION
**Pull request by wodenx**
## Changes
- Adds new token composition utilities (Tokens.fow, Tokens.flowRight, Tokens.meta and Tokens.filter)
- Replaces lodash flow with TOkens.flow in Design api.

## Test Instructions
- Sanity testing on test-site
- Developer should play with new API

## Related Issues
Closes #826

### CODE REVIEW

File: packages/fclasses/README.md [hunk 1]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+```js
+const withComposedToken = asToken(
+  withToken1,
+  withToken2,
+);
+```
+
+However, there are a few key differences:
+
+- Metadata (static properties) attached to a component ar prpagated through
```
Comment (VancheeZze) [inline packages/fclasses/README.md:78]: ar --> are

File: packages/fclasses/README.md [hunk 2]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+aggregated and attached to the target component. See below for some examples.
+
+In addition to a normal HOC, a Token can also be a "filter". A filter is a token
+which, when composed with other tokens, *removes* any which match certain
+criteria. Filters are usually defined to test the metadata attached to other
+tokens. So, for exmple, you can compose a token which removes all 'Color' tokens
+and adds a new one.
+
+> Note that while metadata from all constituent tokens are aggregated and attached
+> to the component to which a composed token is applied, the tomposed token
```
Comment (VancheeZze) [inline packages/fclasses/README.md:102]: tomposed --> composed

**Review by VancheeZze — APPROVED** [review]

File: packages/fclasses/README.md [hunk 3]
```diff
@@ -27,6 +27,156 @@ component changes. Consumers of those components can then sustainably extend and
...
+
+Any HOC can be used as a token, and tokens can be composed using normal
+functional programming paradigms (eg Lodash flow):
+```js
+const withComposedToken = flow(
+  withToken1,
+  withToken2,
+);
+```
+HoweverBodiless provides a token composition utility which adds some
```
Comment (wodenx) [inline packages/fclasses/README.md]:
```suggestion
However, Bodiless provides a token composition utility which adds some
```

File: packages/fclasses/README.md (hunk 2 above)
Comment (wodenx) [reply packages/fclasses/README.md:102]:
```suggestion
> to the component to which a composed token is applied, the composed token
```

File: packages/fclasses/src/Tokens.test.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      const Base = () => <></>;
+      Base.title = 'BaseTitle';
+      const asTest = Tokens.flow(
+        { title: 'Foo' },
+      );
+      const Test = asTest(Base);
+      expect(Test.title).toBe('Foo');
+    });
+
+    it('Adds metaata from nested tokens', () => {
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/fclasses/src/Tokens.test.tsx:111]:
```suggestion
    it('Adds metadata from nested tokens', () => {
```

**Review by VancheeZze — APPROVED** [review]

File: packages/fclasses/README.md [hunk 5]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+const Header1 = asHeader1(H1);  // `H1` is a version of 'h1' stylable with fclasses, see below.
+```
+
+Then
+
+```js
+
+<Header1 /> === <h1 className="text-blue bg-yellow-500 font-bold" />
+
+// The component itself includes aggreaged metadata from all composed tokens...
```
Comment (dewen) [inline packages/fclasses/README.md:156]: aggreaged  => aggregated

File: packages/fclasses/README.md [hunk 6]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
 re-extend their look and feel, with less danger of breakage when the underlying
 component changes.
 
+## Tokens
+
+In bodiless, you implement design tokens as React higer-order components (HOC).
```
Comment (dewen) [inline packages/fclasses/README.md:32]: bodiless => Bodiless

File: packages/fclasses/README.md [hunk 7]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+Applying the HOC to a component is equivalent to styling that component with a
+token:
+
+```js
+const ComponentWithStyles = withMyStyles(Component);
+```
+
+This pattern should be familiar to those who have worked with CSS-in-JS libraries
+like Styled Components or Emotion.
```
Comment (dewen) [inline packages/fclasses/README.md:41]: Emotion => [Emotion](https://emotion.sh/docs/introduction)

File: packages/fclasses/README.md [hunk 8]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+additional functionality:
+
+- The ability to attach metadata to a token.
+- The ability to selectively remove tokens from a composition based on
+  their metadata (or other criteria).
+
+This is intended to promote design-system thinking when defining
+tokens, by encouraging us to think about the structure and organization
+of tokens as we implement them.  It also facilitates implementation of
+tools which allow browsing the design system (eg StorybooK), and eases
```
Comment (dewen) [inline packages/fclasses/README.md:62]: StorybooK => Storybook

File: packages/fclasses/README.md (hunk 1 above)
Comment (dewen) [reply packages/fclasses/README.md:78]: prpagated  => propagated

File: packages/fclasses/README.md [hunk 9]
```diff
@@ -27,6 +27,190 @@ component changes. Consumers of those components can then sustainably extend and
...
+  withToken2,
+);
+```
+
+However, there are a few key differences:
+
+- Metadata (static properties) attached to a component ar prpagated through
+  the chain of HOC's.
+- If you are using Typescript, the type of the parameters is constrained to be an
+  HOC, except...
```
Comment (dewen) [inline packages/fclasses/README.md:81]: "except..."? missing exception details?

File: packages/fclasses/src/Tokens.test.tsx [hunk 10]
```diff
@@ -0,0 +1,214 @@
+import React, { Fragment } from 'react';
```
Comment (dewen) [inline packages/fclasses/src/Tokens.test.tsx:1]: Move this file under 
```packages/fclasses/src/__tests__/```

File: packages/fclasses/src/Tokens.tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      // exclude any token with members, bc we will already be applying the members.
+      ...token.members ? [] : [Object.assign(token, { parents })],
+    ],
+    [] as TokenWithParents<P>[],
+  );
+
+/**
+ * @private
+ *
+ * Generates a token fiter which applies to a token and any of its parents.
```
Comment (dewen) [inline packages/fclasses/src/Tokens.tsx:133]: fiter => filter

File: packages/fclasses/src/Tokens.test.tsx [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+      const asTest = asToken(
+        asFoo,
+        asBaz,
+        { title: 'Test' },
+        meta.term('Type')('Test'),
+      );
+      const Test = asTest(Fragment);
+      expect(Test.title).toBe('Test');
+      expect(Test.categories).toEqual({
+        Type: ['Foo', 'Bar', 'Baz', 'Test'],
```
Comment (dewen) [inline packages/fclasses/src/Tokens.test.tsx:124]: Add a test for meta.cat? Reset meta category

File: packages/fclasses/README.md (hunk 1 above)
Comment (wodenx) [reply packages/fclasses/README.md:78]:
```suggestion
- Metadata (static properties) attached to a component are prppagated through
```

File: packages/fclasses/README.md (hunk 5 above)
Comment (wodenx) [reply packages/fclasses/README.md:156]:
```suggestion
// The component itself includes aggregated metadata from all composed tokens...
```

File: packages/fclasses/README.md (hunk 6 above)
Comment (wodenx) [reply packages/fclasses/README.md:32]:
```suggestion
In Bodiless, you implement design tokens as React higher-order components (HOC).
```

File: packages/fclasses/README.md (hunk 7 above)
Comment (wodenx) [reply packages/fclasses/README.md:41]:
```suggestion
like [Styled Components](https://styled-components.com/) or [Emotion](https://emotion.sh/docs/introduction).
```

File: packages/fclasses/README.md (hunk 8 above)
Comment (wodenx) [reply packages/fclasses/README.md:62]:
```suggestion
tools which allow browsing the design system (eg Storybook), and eases
```

File: packages/fclasses/README.md (hunk 9 above)
Comment (wodenx) [reply packages/fclasses/README.md:81]:
```suggestion
  HOC (or an object specifying metadata, see below).
```

File: packages/fclasses/src/Tokens.tsx (hunk 11 above)
Comment (wodenx) [reply packages/fclasses/src/Tokens.tsx:133]:
```suggestion
 * Generates a token fliter which applies to a token and any of its parents.
```

**Review by dewen — APPROVED** [review]

**Review by VancheeZze — APPROVED** [review]

**Pull request merged**